File: 087_Kth Smallest Element in a BST.cpp
// Kth Smallest Element in a BST
// Find the k-th smallest element using iterative inorder traversal.
#include <bits/stdc++.h>
using namespace std;
struct TreeNode{int val; TreeNode* left; TreeNode* right; TreeNode(int x): val(x), left(nullptr), right(nullptr) {}};

int kthSmallest(TreeNode* root, int k){
    stack<TreeNode*> st; TreeNode* cur = root;
    while(cur || !st.empty()){
        while(cur){ st.push(cur); cur=cur->left; }
        cur = st.top(); st.pop();
        if(--k==0) return cur->val;
        cur = cur->right;
    }
    return -1;
}

int main(){
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(1);
    root->right = new TreeNode(4);
    root->left->right = new TreeNode(2);
    cout<<"kth smallest (k=1): "<<kthSmallest(root,1)<<"\n"; //1
    cout<<"kth smallest (k=3): "<<kthSmallest(root,3)<<"\n"; //3
    return 0;
}

// Time Complexity: O(h + k)
// Space Complexity: O(h)

File: 087_Kth Smallest Element in a BST.py
"""
Kth Smallest Element in a BST
Return the k-th smallest element using inorder traversal.
"""
from typing import Optional

class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val; self.left=left; self.right=right

def kth_smallest(root: Optional[TreeNode], k: int) -> int:
    stack=[]; cur=root
    while cur or stack:
        while cur:
            stack.append(cur); cur=cur.left
        cur = stack.pop(); k-=1
        if k==0: return cur.val
        cur = cur.right
    return -1

if __name__ == '__main__':
    root = TreeNode(3)
    root.left = TreeNode(1)
    root.right = TreeNode(4)
    root.left.right = TreeNode(2)
    print('kth smallest k=1:', kth_smallest(root,1))
    print('kth smallest k=3:', kth_smallest(root,3))

# Time Complexity: O(h + k)
# Space Complexity: O(h)
